### ISSUE TITLE
How to do fine-tuning with ImageAI?

### ISSUE THREAD

**Issue by zlbleues** [OlafenwaMoses/ImageAI#8]
Hi,

Thank you for the great work!

I would like to modify the model resnet50_coco_best_v2.1.0.h5 to modify the weights by adding a sample of data. This sample is not very large.
I would like to refine the object detection on a photo or video, especially looking for cars.

I read the readme and the examples but I did not find a solution to my problem. It's been a while now that I search on the internet but without success. 

Could someone inform me?

thank you in advance

**Comment by OlafenwaMoses** [comment]
Hello,

If you are trying to recognize one or a few distinct object(s), you can use the Custom Object Detection and Custom Objects Video Detection provided in ImageAI. See the links below:
For Images :  https://github.com/OlafenwaMoses/ImageAI/tree/master/imageai/Detection#customdetection

For Video : https://github.com/OlafenwaMoses/ImageAI/blob/master/imageai/Detection/VIDEO.md#customvideodetection

You will find in the above links the different objects the current RetinaNet model file provided can detect. 

In future versions of ImageAI, we will allow training new models for detecting your own list of objects.

**Comment by zlbleues** [comment]
Hello, 
Thank you for your answer. 
I already use the custom object list.

I think I misunderstood. I want to modify the model: resnet50_coco_best_v2.0.1.h5. This is to add new images of cars to improve the detection. How is this possible?

thank you in advance

**Comment by OlafenwaMoses** [comment]
In future versions of ImageAI, we will support training object detection models on new and custom images.

**Comment by OlafenwaMoses** [comment]
Very soon. In maximum of a month's time.

**Comment by OlafenwaMoses** [comment]
@coding-blocks Yes.

**Comment by OlafenwaMoses** [comment]
@zlbleues

**Comment by OlafenwaMoses** [comment]
My team is working on it. Due to other projects that we have at hand, the feature will most likely be available by December this year. However, there will 2 or more releases before then that will come with improvements to existing features and addition of new features and models.

**Comment by drorhilman** [comment]
can you provides a script that show how did you train these models in the first place?

**Comment by yuskey** [comment]
What's the status on this feature? I'd like to train on images with classes outside the 80 included.

**Comment by polytech1** [comment]
> What's the status on this feature? I'd like to train on images with classes outside the 80 included.

Me too

**Comment by Poonamjo** [comment]
Is there any progress on this?

**Comment by Zengdongxu** [comment]
Hello @OlafenwaMoses

I really appreciate your work Imageai, it's short and effective, and I gonna use it in my graduation thesis. But I need to train my own image models outside the 80 classes to perform video detection, is it possible for me to get new release before March? Otherwise I have to chose other libraries?

All the best in your work!

**Comment by alexmotoc** [comment]
Hello, 

I would like to try using custom trained objects for my university project (hands from the VIVA challenge in particular). Any updates on the timeline of this release?

Thanks and keep up the great work guys, this is very intuitive to use!

**Comment by OlafenwaMoses** [comment]
Hi everyone, support for training object detection models on custom dataset/objects will be added later than I initially mentioned. Due to my schedule and work on other projects, I have shifted the window for this to early April.

However, I will update you all soon on a new AI project which will also be open-sourced. The new project will provide more features in recognition, detection and custom models which programmers and developers working with any programming language can use. 

Will drop new updates soon.

**Comment by veesamkrao999** [comment]
> Hello,
> 
> If you are trying to recognize one or a few distinct object(s), you can use the Custom Object Detection and Custom Objects Video Detection provided in ImageAI. See the links below:
> For Images : https://github.com/OlafenwaMoses/ImageAI/tree/master/imageai/Detection#customdetection
> 
> For Video : https://github.com/OlafenwaMoses/ImageAI/blob/master/imageai/Detection/VIDEO.md#customvideodetection
> 
> You will find in the above links the different objects the current RetinaNet model file provided can detect.
> 
> In future versions of ImageAI, we will allow training new models for detecting your own list of objects.

Can we add objects in "object detection model" ?

**Comment by rageshgopan** [comment]
Hello @OlafenwaMoses .. eagerly waiting for training object detection models on custom dataset/objects. Hope you will have a update for us soon 

Thanks in advance !

**Comment by Prati369** [comment]
Hi @OlafenwaMoses . 

Are there any updates on the new AI project? 

Can the model be trained for fish eye images?

**Comment by RomuloUnespian** [comment]
Hi @OlafenwaMoses 

Looking forward for the updates on the AI project!!!

**Comment by xinming31** [comment]
Hi @OlafenwaMoses,
Long time past, we are still waiting.

**Comment by gautamkadian** [comment]
Hi @OlafenwaMoses,

Really appreciate your work on IMAGEAI.

I read out the above message trail, Is there any way to train the same model with car type and etc.
or
Is there any way that i can convert the h5 model into tensorflow model's.

Really appreciate for your reply

Thanks

**Comment by JStuve** [comment]
Excited for this feature!

**Comment by AlenaSuchilkina** [comment]
Good day!
Earlier you wrote that this feature will be released in early April 2019. When will the update be released?
I use your library in my diploma. Many thanks for your work!

**Comment by sathiya1926** [comment]
Hello Moses,

Eagerly waiting for the new project, this will really help us.

**Comment by montycobra** [comment]
Hey imageAI team great job !! While we all wait for this new feature. Can we list down some alternatives here. Don't wish this to be dead end I can see lot of us eagerly awaiting delivery if new feature and it could be helpful to try out this on some alternatives. We can potentially enriched imageAI based on the experience of the alternatives. Win win

**Comment by bmagnetta** [comment]
This will be a great feature! In the meantime I found a work around for a specific case; the YOLOv3 pre-trained model used in ImageAI. There exists code to take YOLOv3 and fine tune it with new data (keras-yolo3). I was able to get an ImageAI object detection example working specifically for cellphone objects. Thought I'd share in case some of you need to implement a feature like this sooner rather than later:  [https://github.com/bmagnetta/cellphonefinder](url)

**Comment by OlafenwaMoses** [comment]
See this article for latest update on this.

https://towardsdatascience.com/the-story-and-future-of-imageai-one-year-anniversary-e63c80f527c8

**Comment by JStuve** [comment]
Following @OlafenwaMoses 

**TLDR: June 15, 2019** 

We get the following related to Custom Modeling: 

* Save custom trained models in full, allowing custom recognition with any Keras model without specifying model type
* Export custom trained model to Tensoflow .pb format

**Comment by nineclicks** [comment]
Forgive me, I am trying to understand. Will this update allow us to train custom objects and do object detection with bounding boxes on custom objects?

**Comment by OlafenwaMoses** [comment]
Hello everyone, we just published a new update **[(v 2.0.3)](https://github.com/OlafenwaMoses/ImageAI)** with the features as mentioned by @JStuve  and more. You can find details on the new features in the repository's description. https://github.com/OlafenwaMoses/ImageAI/blob/master/README.md

Sadly, training of custom models for object detection is yet to be included. We understand this is a highly demanded feature and we are working hard to provide it in the most simple yet customizable implementation possible. Also, delay on this feature has been largely due to the number of projects at hand. 

We have moved the support for training custom object detection models to **July 30, 2019**. All things being equal, we will deliver this feature by this date unfailingly. Thank you all for your patience.

**Comment by Prati369** [comment]
Hi,
Does it support 3D annotation of objects?

**Comment by patrickduhaime** [comment]
> We have moved the support for training custom object detection models to July 30, 2019. All things being equal, we will deliver this feature by this date unfailingly. Thank you all for your patience.

Any update on this ?

Thank you

**Comment by Amir-Omidfar** [comment]
Hi @OlafenwaMoses ! Thanks for your great work!
Just wanted to know if there's any update for custom object detection feature on your platform?

Thanks

**Comment by Gytis123** [comment]
'We have moved the support for training custom object detection models to July 30, 2019. All things being equal, we will deliver this feature by this date unfailingly. Thank you all for your patience.'

Hello

Any updates?

**Comment by OlafenwaMoses** [comment]
Hello everyone, I really appreciate the patience you have all exercised since the beginning of this thread. 

I am glad to announce that training and performing image detection, video detection and video analysis with **YOLOv3** on your **custom image datasets** is now available. **KINDLY READ THE DETAILS BELOW.**

**N.B:**  The new features have been released but not documented yet in **README** and official API documentation. This will will done before the end of the week.
However, we have provided sufficient example codes, example datasets and sample trained models for you to start using this feature. Everything you need to start using the new features are detailed below.


**1) INSTALLATION:**
**ImageAI**  (v2.0.1) is now available on PyPi. To install the latest release, install via the command below.

`pip3 install imageai --upgrade`

**2) ANNOTATING YOUR CUSTOM DATASET**
For custom detection model training, ImageAI uses the Pascal VOC format for image annotation. You can easily annotate your dataset using the tool **LabelIMG** which is available on Windows, Linux and MacOS

https://github.com/tzutalin/labelImg

Also, I will advice that your image dataset contains up to about 200 image samples, from which you can allocate 150 for training and 50 for validation. Your dataset can contain **one or many objects**.

**3)ORGANIZING YOUR CUSTOM DATASET**
Once you have annotated your dataset in the Pascal VOC format, ensure your dataset directory is organized as in the sample below.


`>> train    >> images       >> img_1.jpg
                  >> images       >> img_2.jpg
                  >> images       >> img_3.jpg
                  >> annotations  >> img_1.xml
                  >> annotations  >> img_2.xml
                  >> annotations  >> img_3.xml


>> validation   >> images       >> img_151.jpg
                         >> images       >> img_152.jpg
                         >> images       >> img_153.jpg
                         >> annotations  >> img_151.xml
                         >> annotations  >> img_152.xml
                         >> annotations  >> img_153.xml`


**4) TRAINING ON YOUR CUSTOM DATASET**

You can train your custom detection model completely from scratch or use transfer learning (**recommended** for better accuracy) from a pre-trained YOLOv3 model.  Below is the link to the pre-trained model and also sample datasets you can use.

https://github.com/OlafenwaMoses/ImageAI/releases/tag/essential-v4

The example code to train your object detection model is as linked below after the notes. 

**N.B:** 

- The default batch_size is 4. If you are training with Google Colab, this will be fine. However, I will advice you use a more powerful GPU than the K80 offered by Colab as the higher your batch_size (8, 16), the better the accuracy of your detection model.
- If you experience '_TfDeviceCaptureOp' object has no attribute '_set_device_from_string' in Google Colab, it is due to [a bug in Tensorflow](https://github.com/tensorflow/tensorflow/issues/30728). You can solve this by installing **Tensorflow GPU 1.13.1** `pip3 install tensorflow-gpu==1.13.1`.

https://github.com/OlafenwaMoses/ImageAI/blob/master/examples/custom_detection_train.py

Once your training starts, your models and the _**detection_config.json**_ file you need for performing custom detection will be saved in your **dataset/models** and **dataset/json** folder respectively.


**5) EVALUATING YOUR MODEL's mAP**

After traing your model, you can evaluate the mAP of your models by specifying your desired IoU and Non-maximum suppression values.  The sample codes are linked below.

Single Model Evaluation = https://github.com/OlafenwaMoses/ImageAI/blob/master/examples/custom_detection_single_model_evaluation.py

Multi-Model Evaluation = https://github.com/OlafenwaMoses/ImageAI/blob/master/examples/custom_detection_multi_model_evaluation.py


**6)IMAGE AND VIDEO OBJECT DETECTION WITH YOUR CUSTOM MODEL**

We have provided a number of sample codes for using your custom detection model in detecting objects in images and videos. In addition, we have provided a sample model trained to detect the **Hololens** headset. See the links below for all the resources.

https://github.com/OlafenwaMoses/ImageAI/releases/tag/essential-v4

https://github.com/OlafenwaMoses/ImageAI/blob/master/examples/custom_detection.py

https://github.com/OlafenwaMoses/ImageAI/blob/master/examples/custom_detection_array_input_output.py

https://github.com/OlafenwaMoses/ImageAI/blob/master/examples/custom_detection_extract_objects.py

https://github.com/OlafenwaMoses/ImageAI/blob/master/examples/custom_detection_from_array_extract_objects_array.py

https://github.com/OlafenwaMoses/ImageAI/blob/master/examples/custom_detection_from_file_extract_objects_array.py

https://github.com/OlafenwaMoses/ImageAI/blob/master/examples/custom_detection_video.py

**Comment by OlafenwaMoses** [comment]
GitHub repo documentation updated. Official documentation via https://imageai.readthedocs.io coming  and tutorial coming soon.

**Comment by plenari** [comment]
thanks verymuch

**Comment by OlafenwaMoses** [comment]
Full tutorial on this.

https://medium.com/deepquestai/train-object-detection-ai-with-6-lines-of-code-6d087063f6ff

https://medium.com/deepquestai/object-detection-training-preparing-your-custom-dataset-6248679f0d1d

**Comment by nineclicks** [comment]
Is it possible to keep the classes in the pretrained model? Say I create a model for A, B and C. Then later use that model to train D, how would I create an detection_config.json with A,B,C and D? I am not clear on what the "anchors" are.

**Comment by OlafenwaMoses** [comment]
It will surely be a good idea to keep the classes in the pre-trained model. However in transfer learning, the newly trained model is adjusted to the new training data. The only way to keep the previous classes is to include the images for these classes in the training data.

For the **detection_config.json** file, contents of the JSON file as as detailed below:
- Names of your classes to enable the trained model obtain the corresponding labels for its index representation.
- The anchor box values.  This values are used to define the best fit coordinates for the shape and size of your all objects samples (train data). This values are used by the YOLOv3 architecture to detect objects across 3 scales.  You can learn more via the link below.

https://github.com/pjreddie/darknet/issues/568

**Comment by OlafenwaMoses** [comment]
Hello everyone, please see an urgent bug fix update here #278

**Comment by Eddie-cyberking** [comment]
I got an error while loading custom trained model.
Here is my code
detector = CustomObjectDetection()
detector.setModelTypeAsYOLOv3()
detector.setModelPath(os.path.join(baseFolder, 'models/detection_model-ex-05--loss-20.23.h5'))
detector.setJsonPath(os.path.join(baseFolder, 'json/detection_config.json'))
detector.loadModel()

File "tensorflow-gpu\lib\site-packages\tensorflow\python\framework\ops.py", line 1662, in _create_c_op
    raise ValueError(str(e))

ValueError: Dimension 0 in both shapes must be equal, but are 1 and 18. Shapes are [1,1,1024,855] and [18,1024,1,1]. for 'Assign_360' (op: 'Assign') with input shapes: [1,1,1024,855], [18,1024,1,1].

**Comment by OlafenwaMoses** [comment]
The detection class for _custom detection models_ is different from the _standard object detection_ class. 

Kindly import
`from imageai.Detection.Custom import CustomObjectDetection`
and use 
`CustomObjectDetection`

**Comment by Eddie-cyberking** [comment]
Hi @OlafenwaMoses The problem of #271 due to using the wrong class for custom video detection. But in my case, I am using CustomObjectDetection for image detection. Am I using the class?

**Comment by OlafenwaMoses** [comment]
@Eddie-cyberking have you tried using other models generated during the training? Let me know if you experience the same problem with those too.

**Comment by Eddie-cyberking** [comment]
@OlafenwaMoses It is ok now. The problem is caused by using an unmatched json for the trained model. Thanks for your help.

**Comment by OlafenwaMoses** [comment]
@Eddie-cyberking You are welcome.

**Comment by auphofBSF** [comment]
hello @OlafenwaMoses  I have your hololens training example for Object Detection  working very well thank you.

Now I am trying to setup a custom Object Detection training but with 3 objects Object1, Object2, Object3.  I have tried various combinations of structuring  the training data directories without success.  


could you please supplement your instructions at #issuecomment-516432082  on how to structure with multiple objects

### 3)ORGANIZING YOUR CUSTOM DATASET

`
Once you have annotated your dataset in the Pascal VOC format, ensure your dataset directory is organized as in the sample below.


train >> images >> img_1.jpg
train >> images >> img_2.jpg
train >> images >> img_3.jpg
train >> annotations >> img_1.xml
train >> annotations >> img_2.xml
train >> annotations >> img_3.xml

validation >> images >> img_151.jpg
validation >> images >> img_152.jpg
validation >> images >> img_153.jpg
validation >> annotations >> img_151.xml
validation >> annotations >> img_152.xml
validation >> annotations >> img_153.xml``

**Comment by OlafenwaMoses** [comment]
All you need to do is keep all images of all objects in the same folder as well as their corresponding annotation XMLs in the same folder. Then set your **object_names_array** in the **.setTrainConfig()** function to the names of the unique objects as seen in the example below.

trainer.setTrainConfig(object_names_array=["hololens", "google-glass", "oculus"], ........)

**Comment by OlafenwaMoses** [comment]
@auphofBSF I have updated the code and README documentation to clarify this. See the link below.

https://github.com/OlafenwaMoses/ImageAI/blob/master/imageai/Detection/Custom/CUSTOMDETECTIONTRAINING.md

**Comment by auphofBSF** [comment]
> @auphofBSF I have updated the code and README documentation to clarify this. See the link below.
> 
> https://github.com/OlafenwaMoses/ImageAI/blob/master/imageai/Detection/Custom/CUSTOMDETECTIONTRAINING.md#trainingdataset

Thank you very much  @OlafenwaMoses , you have got me onto the right track and all appears to be working well.
I have further to your doc update submitted PR #295 to further clarify directory structure for multiple object learning

**Comment by OlafenwaMoses** [comment]
> > @auphofBSF I have updated the code and README documentation to clarify this. See the link below.
> > https://github.com/OlafenwaMoses/ImageAI/blob/master/imageai/Detection/Custom/CUSTOMDETECTIONTRAINING.md#trainingdataset
> 
> Thank you very much @OlafenwaMoses , you have got me onto the right track and all appears to be working well.
> I have further to your doc update submitted PR #295 to further clarify directory structure for multiple object learning

You are most welcome. I have made some comment on the PR.

**Comment by chldjwls123** [comment]
when i run custom training code, error comes up like below one.

"AttributeError: module '_pywrap_tensorflow_internal' has no attribute 'CudaSupportsHalfMatMulAndConv'"

how can i fix it??

**Comment by chldjwls123** [comment]
When i was running train-code, I got this error message:

Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "D:\pycharm\PyCharm Community Edition 2019.2.1\helpers\pydev\_pydev_bundle\pydev_umd.py", line 197, in runfile
    pydev_imports.execfile(filename, global_vars, local_vars)  # execute the script
  File "D:\pycharm\PyCharm Community Edition 2019.2.1\helpers\pydev\_pydev_imps\_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "C:/Users/User/PycharmProjects/CAPSTONE/detection_training.py", line 10, in <module>
    trainer.trainModel()
  File "C:\Users\User\PycharmProjects\CAPSTONE\venv\lib\site-packages\imageai\Detection\Custom\__init__.py", line 301, in trainModel
    max_queue_size=8
  File "C:\Users\User\PycharmProjects\CAPSTONE\venv\lib\site-packages\keras\legacy\interfaces.py", line 91, in wrapper
    return func(*args, **kwargs)
  File "C:\Users\User\PycharmProjects\CAPSTONE\venv\lib\site-packages\keras\engine\training.py", line 1658, in fit_generator
    initial_epoch=initial_epoch)
  File "C:\Users\User\PycharmProjects\CAPSTONE\venv\lib\site-packages\keras\engine\training_generator.py", line 215, in fit_generator
    class_weight=class_weight)
  File "C:\Users\User\PycharmProjects\CAPSTONE\venv\lib\site-packages\keras\engine\training.py", line 1449, in train_on_batch
    outputs = self.train_function(ins)
  File "C:\Users\User\PycharmProjects\CAPSTONE\venv\lib\site-packages\keras\backend\tensorflow_backend.py", line 2979, in __call__
    return self._call(inputs)
  File "C:\Users\User\PycharmProjects\CAPSTONE\venv\lib\site-packages\keras\backend\tensorflow_backend.py", line 2937, in _call
    fetched = self._callable_fn(*array_vals)
  File "C:\Users\User\PycharmProjects\CAPSTONE\venv\lib\site-packages\tensorflow\python\client\session.py", line 1439, in __call__
    run_metadata_ptr)
  File "C:\Users\User\PycharmProjects\CAPSTONE\venv\lib\site-packages\tensorflow\python\framework\errors_impl.py", line 528, in __exit__
    c_api.TF_GetCode(self.status.status))
tensorflow.python.framework.errors_impl.ResourceExhaustedError: OOM when allocating tensor with shape[2,26,26,512] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
    [[{{node replica_1/model_1/leaky_42/LeakyRelu}}]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.
    [[{{node replica_0/model_1/bnorm_62/FusedBatchNorm}}]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

How can i fix it...?

**Comment by Thanakorn-VK** [comment]
I got an error please help!

Generating anchor boxes for training images and annotation...
Average IOU for 9 anchors: 0.95
Anchor Boxes generated.
Detection configuration saved in  nice/json/detection_config.json
Some labels have no annotations! Please revise the list of labels in your configuration.
Training on: 	None
Training with Batch Size:  4
Number of Experiments:  100
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-8-9805d7882459> in <module>()
      5 trainer.setDataDirectory(data_directory="nice")
      6 trainer.setTrainConfig(object_names_array=["nappacabbage"], batch_size=4, num_experiments=100, train_from_pretrained_model="pretrained-yolov3.h5")
----> 7 trainer.trainModel()

1 frames
/usr/local/lib/python3.6/dist-packages/imageai/Detection/Custom/generator.py in __init__(self, instances, anchors, labels, downsample, max_box_per_image, batch_size, min_net_size, max_net_size, shuffle, jitter, norm)
     34         self.net_w              = 416
     35 
---> 36         if shuffle: np.random.shuffle(self.instances)
     37 
     38     def __len__(self):

mtrand.pyx in numpy.random.mtrand.RandomState.shuffle()

TypeError: object of type 'NoneType' has no len()

**Comment by anujapinge** [comment]
Hello, I am trying to perform transfer learning on the yolov3.  Why it is recommended to have 200 images of each object to be trained? And also can I include as many number of classes as I want?
